crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4301,6 +4403,18 @@ impl<'db> TypeInferenceBuilder<'db> {
         let mut call_arguments = Self::parse_arguments(arguments);
         let callable_type = self.infer_expression(func);
 
+        if let Type::FunctionLiteral(function) = callable_type {
+            // Make sure that the `function.definition` is only called when the function is defined
+            // in the same file as the one we're currently inferring the types for. This is because
+            // the `definition` method accesses the semantic index, which could create a
+            // cross-module AST dependency.
+            if function.body_scope(self.db()).file(self.db()) == self.file()
```

Comment:
minor nit: given that we now do this check a couple places, it could be a method `FunctionType::defined_in_file(db, file)`?

Reply:
Are you referring to the `function.body_scope(self.db()).file(self.db())` part or the entire equality check? I'm guessing it's the former as I couldn't find the latter part.

Reply:
I meant the entire equality check. Don't we do it both here and in the assertion you added in `FunctionType::node`?

I mean, extracting a `FunctionType::file()` method isn't necessarily a bade idea either...

Reply:
> I mean, extracting a `FunctionType::file()` method isn't necessarily a bade idea either...

Yeah, I've added this instead of the equality check method.